paddle/phi/kernels/fusion/cutlass/moe_kernel_impl.h

```diff
@@ -0,0 +1,782 @@
+                        const int inter_size,
+                        const int num_experts,
+                        const int k,
+                        const int batch_size,
+                        const int max_seq_len) {
+  const int buf_size =
+      pad_to_multiple_of_16(num_experts * batch_size * k * hidden_size);
+  const int interbuf_size =
+      pad_to_multiple_of_16(num_experts * batch_size * k * inter_size);
+  const int padded_experts = pad_to_multiple_of_16(num_experts);
+  const int num_moe_inputs = pad_to_multiple_of_16(num_experts * num_rows);
+  int padded_num_moe_inputs = num_experts * batch_size * max_seq_len;
+
+  size_t total_ws_bytes = sizeof(int) * num_moe_inputs;   // source_rows_
+  total_ws_bytes += sizeof(int) * padded_num_moe_inputs;  // padded_source_rows_
+  total_ws_bytes += sizeof(T) * padded_num_moe_inputs;  // padded_expert_scales_
+  total_ws_bytes += sizeof(int) * padded_num_moe_inputs;  // permuted_rows_
+  total_ws_bytes += sizeof(int) * num_experts * k;        // permuted_experts_
+  total_ws_bytes += buf_size * sizeof(T);                 // permuted_data_
+  total_ws_bytes +=
+      padded_experts * sizeof(int64_t);  // Hold total_rows_before_expert_
+
+  total_ws_bytes += sizeof(T) * num_moe_inputs;         // attr_mask: [e, n]
+  total_ws_bytes += sizeof(T) * padded_num_moe_inputs;  // sorted_softmax_output
+
+  const int bytes_for_fc1_result = interbuf_size * sizeof(T);
+  const int sorter_ws_size_bytes =
+      pad_to_multiple_of_16(sorter_.getWorkspaceSize(num_experts * k));
+  sorter_.update_num_experts(k);
+
+  int bytes_for_intermediate_and_sorting = bytes_for_fc1_result;
+  if (sorter_ws_size_bytes > bytes_for_fc1_result) {
+    int remaining_bytes =
+        pad_to_multiple_of_16(sorter_ws_size_bytes - bytes_for_fc1_result);
+    bytes_for_intermediate_and_sorting += remaining_bytes;
+  }
+
+  total_ws_bytes +=
+      bytes_for_intermediate_and_sorting;  // intermediate (fc1) output + cub
+                                           // sorting workspace
+  return total_ws_bytes;
+}
+
+// --------      initialize_expert_choice_route_kernel      -------- //
+template <typename T>
+__global__ void initialize_expert_choice_route_kernel(
+    int* expert_for_source_row,
+    int* source_row,
+    int* expanded_source_row_to_expanded_dest_row,
+    int64_t* total_rows_before_expert,
+    T* attr_mask,
+    const int cols,
+    const int k,
+    const int batch_size) {
+  int start = cols * blockIdx.x;
+
+  for (int i = threadIdx.x; i < cols; i += blockDim.x) {
+    expert_for_source_row[start + i] = blockIdx.x;
+    source_row[start + i] = start + i;
+    expanded_source_row_to_expanded_dest_row[start + i] = -1;
+    attr_mask[start + i] = (T)1.0f;
+  }
+  if (threadIdx.x == 0) {
+    total_rows_before_expert[blockIdx.x] = batch_size * k * (blockIdx.x + 1);
+  }
+}
+
+// --------      softmax_kernel      -------- //
+template <int ITEMS_PER_THREAD, typename T>
+__global__ void softmax_kernel_v4(
```

Comment:
About the naming, there is no v1, v2 and v3. And add comments for the difference between v4 and v5